lib/services/consumers/application.service.js

```diff
@@ -1,30 +1,30 @@
-'use strict';
-
+const uuidv4 = require('uuid/v4');
+const { validate } = require('../../../lib/schemas');
 const applicationDao = require('./application.dao.js');
 const config = require('../../config');
 const utils = require('../utils');
-const uuidv4 = require('uuid/v4');
+
+const SCHEMA = 'http://express-gateway.io/models/applications.json';
 
 const s = {};
 
 s.insert = function (_app, userId) {
-  let app;
   try {
-    app = validateAndCreateApp(_app, userId);
+    return validateAndCreateApp(_app, userId)
+      .then((app) => Promise.all([app, applicationDao.insert(app)]))
```

Comment:
you lost me there. what is `app => Promise.all([app, ...])` ? why app is a promise ?

Reply:
`app` is not a Promise, it's a simple value being passed to the next function in the chain.

This is the usual way to pass values down to a promise chain. This improves the locality of the task because you do not have to rely on global values spread in the test file.

So basically with this trick you can do this

```javascript
getUser('Vincenzo')
  .then((user)=>Promise.all([user, getCredentials(user)]))
  .then(([user, credentials]) => {/*…stuff…*/})
```

instead of 

```javascript
let theUser = undefined;

getUser('Vincenzo')
  .then((user)=>{ theUser = user; return getCredentials(user)})
  .then((credentials) => {/*…stuff…*/})
```

It is basically the Promise version of [async.waterfall](https://caolan.github.io/async/docs.html#waterfall)

Does that clarify?

Reply:
ok, I get it. dao.insert for some reason returns `true` instead of inserted app

can you check is it really have to return true\not-true or it can just return the app ?

Reply:
I think it should technically be possible.

The only thing I should figure out is why [this check](https://github.com/ExpressGateway/express-gateway/blob/master/lib/services/consumers/application.dao.js#L27) is being performed (and the result returned) on the DAO object once the object has been inserted and more importantly who's relying on it.

Reply:
be careful with `.then(res => res.every(val => val));` last time I was trying to remove it everything blew up  

up to you, I'm ok with refactoring it in another task